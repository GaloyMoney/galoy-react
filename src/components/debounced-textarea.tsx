import { ChangeEvent, useState, memo, useEffect } from "react"
import { useDebouncedCallback } from "use-debounce"

export type OnTextValueChange = (value: string) => void

type Props = {
  onChange?: OnTextValueChange
  onDebouncedChange?: OnTextValueChange
  initValue?: string
  debounceDelay?: number
  [prop: string]: unknown
}

type InputObject = {
  value: string
  debouncedValue?: string
  typing: boolean
}

const DebouncedTextareaComponent = ({
  onChange,
  onDebouncedChange,
  initValue,
  debounceDelay = 1500,

  ...textAreaProps
}: Props) => {
  const [input, setInput] = useState<InputObject>({
    value: initValue ?? "",
    typing: false,
  })

  const setDebouncedInputValue = useDebouncedCallback((debouncedValue) => {
    setInput((currInput) => ({ ...currInput, debouncedValue, typing: false }))
  }, debounceDelay)

  useEffect(() => {
    if (input.typing) {
      setDebouncedInputValue(input.value)
    }
    return () => setDebouncedInputValue.cancel()
  }, [setDebouncedInputValue, input.typing, input.value])

  useEffect(() => {
    if (onDebouncedChange && input.debouncedValue !== undefined) {
      onDebouncedChange(input.debouncedValue)
    }
  }, [onDebouncedChange, input.debouncedValue])

  const handleOnChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
    const newValue = event.target.value
    if (onChange) {
      onChange(newValue)
    }
    setInput({ value: newValue, typing: true })
  }

  return <textarea value={input.value} onChange={handleOnChange} {...textAreaProps} />
}

export const DebouncedTextarea = memo(DebouncedTextareaComponent)
