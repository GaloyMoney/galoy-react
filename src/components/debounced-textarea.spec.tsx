import { render } from "@testing-library/react"

import { DebouncedTextarea } from "./debounced-textarea"

describe("DebouncedTextarea", () => {
  it("renders and matches snapshot", () => {
    const { asFragment } = render(<DebouncedTextarea />)
    expect(asFragment()).toMatchSnapshot()
  })
})
