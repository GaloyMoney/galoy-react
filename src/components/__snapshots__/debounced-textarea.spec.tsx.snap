// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebouncedTextarea renders and matches snapshot 1`] = `
<DocumentFragment>
  <textarea />
</DocumentFragment>
`;
